### ISSUE TITLE
Bug in F* : Able to Make a GTot into a Tot

### ISSUE THREAD

**Issue by jaybosamiya** [FStarLang/FStar#1750]
```
module Bug

type eff_tag =
  | G
  | T

let eff (t:eff_tag) (b:Type) =
  match t with
  | T -> unit -> Tot b
  | G -> unit -> GTot b

let eff_to_tot (x:eff T 'a) : Tot 'a = x ()

let gtot_to_eff #a #b (f:(x:a -> GTot (b x))) : x:a -> eff G (b x) = fun x () -> f x

val bug : #t:Type0 -> (x:Ghost.erased t) -> Tot (y:t{y == Ghost.reveal x})
let bug #t x =
  let foo #e (x:Ghost.erased t) : eff e t = (gtot_to_eff Ghost.reveal) x in
  eff_to_tot (foo x)
```

**Comment by nikswamy** [comment]
Thanks @jaybosamiya !

Been discussing this with @jaybosamiya @R1kM and @tahina-pro 

This bug breaks the soundness of erasure and hence the soundness of extraction. 

A smaller variant of it is:

```
let t_f = bool -> Tot bool
let t_g = bool -> GTot bool
let test (b:bool) (f:'a{'a == t_f}) (g:'b{'b == t_g}) =
  if b then f else g
```

What's happening is that the SMT encoding isn't distinguishing between Tot and GTot functions, in two regards:

1. Encodings of `(a -> Tot b)` and `(a -> GTot b)` are hash-consed to the same arrow Tm_arrow constructor in the SMT encoding. An upcoming fix distinguishes these cases by encoding the latter to an instance of `Tm_ghost_arrow …` 

2. The encoding of the HasType relation for Tm_arrow includes both an introduction and an elimination rule. So, even by fixing 1 and encoding them to the distinct type constructors, it is easy for SMT to prove that `HasType f (a -> Tot b) <==> HasType f (a -> GTot b)`.  To fix this, we consider removing the introduction rule for arrow types in the SMT encoding. And @R1kM had a nice suggestion that we might only remove this introduction rule for Tot arrows, retaining it for GTot arrows, since GTot introduction is less demanding.

Going to try this patch for regressions now.

**Comment by nikswamy** [comment]
@mtzguido @aseemr : would be great if you guys can take a look at the issue and my patch too

**Comment by nikswamy** [comment]
ha. *  Failure - There were errors in: FStar.Set.fst, FStar.PredicateExtensionality.fst, FStar.SquashProperties.fst, FStar.Map.fst, FStar.DependentMap.fst, FStar.List.Tot.Base.fst

Looking at it

**Comment by aseemr** [comment]
Thanks Nik, I will take a look. But instead of making the encoding less precise by removing the intro, why not add the `Tot` and `GTot` info to the encoding?

**Comment by nikswamy** [comment]
This is indeed what I'm trying now in https://github.com/FStarLang/FStar/commit/0de820d5e12950343d1ab907c80f42d84c354b09

But, I'm uncovering other problems in the SMT encoding related to the encoding the types of partial applications.

**Comment by mtzguido** [comment]
Playing with this I minimized the (first part of the) bug to:
```fstar
let bug1 (a:Type) = assert ((unit -> Tot a) == (unit -> GTot a))                                     
```
which correctly fails after Nik's patch. I tried to see if tactics had any relevancy here, and found that this:
```fstar
let bug2 (a:Type) = assert ((unit -> Tot a) == (unit -> GTot a)) by (trefl ())                       
```
**succeeds** (without SMT) since there's a long-standing bug in `solve_c` which solves `Tot a =? GTot a` (but not `GTot a =? Tot a`, probably sub-effecting caused the confusion). The fix is easy (see `guido_misc` / bac54f3f5c4a48b36a8e0565fa009a670096cd37). Mostly mentioning this now since I ran into a couple of regressions as well, so we fix them in one pass :).

I'll provide comments on the SMT aspect soon.
